rmw_dds_common/test/test_graph_cache.cpp

```diff
@@ -0,0 +1,1113 @@
+    // topic3
+    {"reader7", "topic3", "Float", true},
+  });
+
+  // Check graph state.
+  check_results(
+    graph_cache,
+    {},
+  {
+    {"topic1", {"Str"}},
+    {"topic2", {"Str"}},
+    {"topic4", {"Int"}},
+  });
+  check_results_by_node(graph_cache, "ns", "name");
+  check_results_by_topic(graph_cache, "topic1", 1, 1);
+  check_results_by_topic(graph_cache, "topic2", 1, 0);
+  check_results_by_topic(graph_cache, "topic3", 0, 0);
+  check_results_by_topic(graph_cache, "topic4", 0, 1);
+
+  // Remove the remaining readers and writers.
+  remove_entities(
+    graph_cache,
+  {
+    // topic1
+    {"reader1", "topic1", "Str", true},
+    {"writer1", "topic1", "Str", false},
+    // topic2
+    {"reader5", "topic2", "Str", true},
+    // topic4
+    {"writer8", "topic4", "Int", false},
+  });
+
+  // Check graph state.
+  check_results(graph_cache);
+  check_results_by_node(graph_cache, "ns", "name");
+  check_results_by_topic(graph_cache, "topic1");
+  check_results_by_topic(graph_cache, "topic2");
+  check_results_by_topic(graph_cache, "topic3");
+  check_results_by_topic(graph_cache, "topic4");
+}
+
+void
+add_participants(
+  GraphCache & graph_cache,
+  const std::vector<std::string> & gids)
+{
+  for (const auto & gid : gids) {
+    graph_cache.add_participant(gid_from_string(gid));
+  }
+}
+
+void
+remove_participants(
+  GraphCache & graph_cache,
+  const std::vector<std::string> & gids)
+{
+  for (const auto & gid : gids) {
+    graph_cache.remove_participant(gid_from_string(gid));
+  }
+}
+
+rmw_dds_common::msg::Gid
+gid_msg_from_string(const std::string & str)
+{
+  rmw_dds_common::msg::Gid gid;
+  gid.data = {0};
+  EXPECT_LT(str.size(), RMW_GID_STORAGE_SIZE);
+  std::strncpy(
+    reinterpret_cast<char *>(gid.data.data()),
+    str.c_str(),
+    RMW_GID_STORAGE_SIZE);
+  return gid;
+}
+
+struct NodeInfo
+{
+  std::string gid;
+  std::string namespace_;
+  std::string name;
+};
+
+using NodeInfoVec = std::vector<NodeInfo>;
+
+rmw_dds_common::msg::ParticipantEntitiesInfo
+add_nodes(
+  GraphCache & graph_cache,
+  const NodeInfoVec & node_info)
+{
+  rmw_dds_common::msg::ParticipantEntitiesInfo msg;
+  for (const auto & elem : node_info) {
+    msg = graph_cache.add_node(
+      gid_from_string(elem.gid),
+      elem.name,
+      elem.namespace_);
+  }
+  return msg;
+}
+
+rmw_dds_common::msg::ParticipantEntitiesInfo
+remove_nodes(
+  GraphCache & graph_cache,
+  const NodeInfoVec & node_info)
+{
+  rmw_dds_common::msg::ParticipantEntitiesInfo msg;
+  for (const auto & elem : node_info) {
+    msg = graph_cache.remove_node(
+      gid_from_string(elem.gid),
+      elem.name,
+      elem.namespace_);
+  }
+  return msg;
+}
+
+struct NodeEntitiesInfo
+{
+  std::string namespace_;
+  std::string name;
+  std::vector<std::string> reader_gids;
+  std::vector<std::string> writer_gids;
+};
+
+struct ParticipantEntitiesInfo
+{
+  std::string gid;
+  std::vector<NodeEntitiesInfo> node_entities_info_seq;
```

Comment:
@ivanpauno nit^2: why the trailing `_seq`?

Reply:
From sequence, to imitate how the similar field is named in the message definition.